### DISCUSSION TITLE
Starting an embedded IPython console from textual

### DISCUSSION THREAD

**Discussion by blob42** [Textualize/textual#2741] [Q&A]
Hi all, 

First thanks to the Textual team for this amazing project.

I am trying to display an embedded IPython console from within a textual app namespace. I came up with the following hack which works fine. However I think when there aare timer based updates like cursor blink or a spinner it keeps "leaking" from the render loop into the ipython console. 

https://github.com/Textualize/textual/assets/210457/bafbe466-5d3a-4228-9f2f-8a538d9992ea

Is there a clean way to achieve what I want ?

```python
    async def action_dev_console(self) -> None:
        """Start the dev console"""
        assert self._driver is not None

        if WINDOWS:
            self.post_message(
                LogMessage.warning(
                    "dev console is not working yet in windows."))
            return

        self.set_focus(None)
        self.refresh()
        await _asyncio.sleep(0.3)

        self._driver.stop_application_mode()
        self._driver.flush()

        stdout_orig = sys.stdout
        stderr_orig = sys.stderr
        driver = t.cast('LinuxDriver', self._driver)
        sys.stdout = sys.stderr = driver._file

        self._embed(colors="neutral")
        sys.stdout = stdout_orig
        sys.stderr = stderr_orig
        driver.start_application_mode()
```

**Comment by TomJGooding** [comment]
I've only seen these suggested in other similar discussions and not tested personally, but you might find this [textual-terminal](https://github.com/mitosch/textual-terminal) package useful and the [Pyte example](https://github.com/selectel/pyte/blob/master/examples/terminal_emulator.py) it is based on.

**Comment by blob42** [comment]
Thanks @TomJGooding for these examples. It might come in handy in the future. I came across something similar already when going through textual issues. 

My situation is a bit different. I want to embed the ipython console in the namespace of the a running Textual app so I cannot fork a new process.

I have an almost working solution. If only there was a way to somehow pause the render loop or the timers driving the updates. For now I need to ensure no animation loop or focused input is active but as soon as I do something from the ipython console that changes the state of the app the render loop draws over the ipython console.

**Reply by dbieber** [reply]
Hi @blob42 - did you get this working?

**Reply by blob42** [reply]
Yes with a very hacky way but I could achieve exactly what you want, dropping the user into an ipython shell from a running textual app. Checkout  https://github.com/blob42/Instrukt the interesting code is in https://github.com/blob42/Instrukt/blob/c15cde6366922af3dc250f9b103b3bf56c0cc516/instrukt/app.py#L396C1-L396C1  and subprocess.py

**Reply by newt32389** [reply]
Was anyone able to create a standalone textual App to do this? Where it's only point is to offer an ipython repl with access to the underlying globals.

I am trying to write an application with two panes. One is a ipython repl and another is a table view of some data that is updated based on the repl interaction
